File: union-find/python/UnionFind.py
#!/usr/local/bin/python3
# -*- coding: utf-8 -*-

class UnionFind(object):

	def __init__(self, N):
		self.N = N

	def union(self, p, q):
		raise NotImplementedError

	def connected(self, p, q):
		raise NotImplementedError

	def __repr__(self):
		raise NotImplementedError

def main(cls):
	N = int(input("Enter N: "))
	uf = cls(N)

	while True:
		nodes = [int(i) for i in input().split()]

		if not nodes:
			break
		else:
			p,q = nodes

		if uf.connected(p, q):
			print("{} and {} are already connected.".format(p, q))
		else:
			uf.union(p, q)
			print("{} -> {}".format(p, q))
		print(uf)

File: union-find/java/src/UnionFind.java
/**
 * Created by petergoldsborough on 08/12/15.
 */

public abstract class UnionFind
{
    UnionFind(int N)
    {
        _N = N;

        _id = new int[N];

        for (int i = 0; i< N; ++i)
        {
            _id[i] = i;
        }
    }

    abstract public void union(int p, int q);

    abstract public boolean connected(int p, int q);

    public int N()
    {
        return _N;
    }

    public void print()
    {
        for (int i : _id) System.out.printf("%d ", i);

        System.out.println();
    }

    protected int[] _id;

    protected int _N;
}
